File: fm-domain/src/main/java/io/zeitwert/fm/dms/adapter/api/rest/DocumentContentController.java
package io.zeitwert.fm.dms.adapter.api.rest;

import io.zeitwert.fm.app.model.SessionContextFM;
import io.zeitwert.fm.dms.model.ObjDocument;
import io.zeitwert.fm.dms.model.ObjDocumentRepository;
import io.zeitwert.fm.dms.model.enums.CodeContentType;
import org.springframework.http.ContentDisposition;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

@RestController("documentContentController")
@RequestMapping("/rest/dms/documents")
public class DocumentContentController {

	private final ObjDocumentRepository documentRepository;
	private final SessionContextFM sessionContext;

	public DocumentContentController(ObjDocumentRepository documentRepository, SessionContextFM sessionContext) {
		this.documentRepository = documentRepository;
		this.sessionContext = sessionContext;
	}

	@RequestMapping(value = "/{documentId}/content", method = RequestMethod.GET)
	public ResponseEntity<byte[]> getContent(@PathVariable Integer documentId) {
		if (documentId == null) {
			return ResponseEntity.notFound().build();
		}
		ObjDocument document = this.documentRepository.get(documentId);
		CodeContentType contentType = document.getContentType();
		if (contentType == null) {
			return ResponseEntity.noContent().build();
		}
		ContentDisposition contentDisposition = ContentDisposition.builder("inline").filename(document.getName()).build();
		HttpHeaders headers = new HttpHeaders();
		headers.setContentDisposition(contentDisposition);
		return ResponseEntity.ok().contentType(contentType.getMediaType()).headers(headers).body(document.getContent());
	}

	@RequestMapping(value = "/{documentId}/content", method = RequestMethod.POST)
	public ResponseEntity<Void> setContent(@PathVariable Integer documentId,
																				 @RequestParam("file") MultipartFile file) {
		try {
			CodeContentType contentType = CodeContentType.getContentType(file.getContentType(),
					file.getOriginalFilename());
			if (contentType == null) {
				return ResponseEntity.badRequest().body(null);
			}
			ObjDocument document = this.documentRepository.load(documentId);
			document.storeContent(contentType, file.getBytes(), sessionContext.getUser().getId(), sessionContext.getCurrentTime());
		} catch (IOException e) {
			e.printStackTrace();
			return ResponseEntity.internalServerError().body(null);
		}
		return ResponseEntity.ok().body(null);
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/dms/adapter/rest/DocumentContentController.kt
package io.zeitwert.fm.dms.adapter.rest

import io.zeitwert.app.session.model.SessionContext
import io.zeitwert.fm.dms.model.ObjDocumentRepository
import io.zeitwert.fm.dms.model.enums.CodeContentType.Enumeration.getContentType
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.Resource
import org.springframework.http.ContentDisposition
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@RestController("documentContentController")
@RequestMapping("/rest/dms/documents")
class DocumentContentController(
	private val documentRepository: ObjDocumentRepository,
	private val sessionContext: SessionContext,
) {

	@Value("classpath:templates/missing.jpg")
	lateinit var missingImage: Resource

	@GetMapping(value = ["/{documentId}/content"])
	fun getContent(
		@PathVariable documentId: Int,
	): ResponseEntity<ByteArray> {
		val document = documentRepository.get(documentId)
		val contentType = document.contentType
		val content = if (contentType == null) {
			missingImage.inputStream.readAllBytes()
		} else {
			document.content
		}
		val contentDisposition = ContentDisposition.builder("inline").filename(document.name).build()
		val headers = HttpHeaders()
		headers.contentDisposition = contentDisposition
		return ResponseEntity
			.ok()
			.contentType(contentType?.getMediaType() ?: MediaType.APPLICATION_OCTET_STREAM)
			.headers(headers)
			.body(content)
	}

	@PostMapping(value = ["/{documentId}/content"])
	fun setContent(
		@PathVariable documentId: Int,
		@RequestParam("file") file: MultipartFile,
	): ResponseEntity<Void> {
		try {
			val contentType = getContentType(
				file.contentType,
				file.originalFilename,
			)
			if (contentType == null) {
				return ResponseEntity.badRequest().body(null)
			}
			val document = documentRepository.load(documentId)
			document.storeContent(contentType, file.bytes, sessionContext.userId, sessionContext.currentTime)
		} catch (e: IOException) {
			e.printStackTrace()
			return ResponseEntity.internalServerError().body(null)
		}
		return ResponseEntity.ok().body(null)
	}

}
